File: src/python/oop/ElevatorSystem/elevator/default_dispatch_strategy.py
import random

from .dispatch_strategy_base import DispatchStrategyBase
from .elevator import Elevator
from .direction import Direction


class DefaultDispatchStrategy(DispatchStrategyBase):
    def find_elevator(self, elevators: list[Elevator], current_floor: int, direction: Direction):
        for elevator in elevators:
            if current_floor not in elevator.floors:
                continue

            status = elevator.get_status()
            if status.direction == Direction.IDLE:
                return elevator

            if status.direction == direction \
                and ((status.direction == Direction.DOWN and status.current_floor > current_floor) \
                or (status.direction == Direction.UP and status.current_floor < current_floor)):
                return elevator

        return elevators[random.randint(0, len(elevators) - 1)]

File: src/csharp/oop/ElevatorSystem/Elevator/DefaultDispatchStrategy.cs
namespace ElevatorSystem.Elevator;

public class DefaultDispatchStrategy : DispatchStrategyBase
{
    public override ElevatorCabin? FindElevator(IList<ElevatorCabin> elevators, int floor, Direction direction)
    {
        foreach (var elevator in elevators)
        {
            if (!elevator.Floors.Contains(floor))
            {
                continue;
            }

            if (elevator.Status.Direction == Direction.Idle)
            {
                return elevator;
            }

            if (elevator.Status.Direction == direction
                || ((elevator.Status.Direction == Direction.Down || elevator.Status.CurrentFloor > floor)
                    && (elevator.Status.Direction == Direction.Up || elevator.Status.CurrentFloor < floor)))
            {
                return elevator;
            }
        }

        return elevators[Random.Shared.Next(0, elevators.Count)];
    }
}
